### DISCUSSION TITLE
Mimic Unity child container behavior

### DISCUSSION THREAD

**Discussion by djymdmannen** [dadhi/DryIoc#579] [Q&A]
I'm trying out DryIoc as a replacement for Unity, and I can't quite figure out how to mimic the behavior of child containers in Unity.

In Unity the registrations actively flow from the parent container to any child containers, but not from the child containers to the parent container.

In DryIoc I've so far only been able to either clone the registrations or fully share the registrations. (parent <-> children). Anyone got any pointers for me?

**Comment by dadhi** [comment]
@djymdmannen Hi there.

I am not familiar with the Unity behavior. But I would imagine that cloning of the registrations from the parent provides the better stability and isolation for the child. If you need to add the new registration to the parent and make it visible in the children, you may try to either rethink it :-) and implement this dynamic behavior via other abstraction (and avoid DI completely). Or you may do some clever things like holding the references to child containers near the parent and duplicate the registrations into them.

Basically, it all depends on the use-case. I can help if you provide more details or better a sample code.

**Comment by djymdmannen** [comment]
Thanks for a quick answer!

I agree. For the sake of stability and determinism I'd prefer that registrations are done fully in the parent container before the children are created. But I need to address that as a separate issue.

There's not much sample code to show, but maybe these two tests can illustrate what I'm after:
```
public class ChildContainerBehaviorTests
{
    public sealed record SomeType1();

    [Fact]
    public void Registrations_In_Child_Does_Not_Flow_To_Parent()
    {
        IContainer parent = new Container();
        IContainer child = parent.CreateChild();

        child.Register<SomeType1>();

        Assert.False(parent.IsRegistered<SomeType1>());
        Assert.True(child.IsRegistered<SomeType1>());
    }

    [Fact]
    public void Registrations_In_Parent_Flows_To_Child()
    {
        IContainer parent = new Container();
        IContainer child = parent.CreateChild();

        parent.Register<SomeType1>();

        Assert.True(parent.IsRegistered<SomeType1>());
        Assert.True(child.IsRegistered<SomeType1>());
    }
}
```
The last assert in the last test fails. And of course, If I use `RegistrySharing.Share` then the first test fails instead.

**Reply by dadhi** [reply]
@djymdmannen 

Yeah, your tests display exactly what I was talking about, so no surprise here.
As I said, you may either design so, that you do all the registrations into the parent before creating the child. 
Or you may do something pretty straightforward like this:

```cs
using System;
using DryIoc;

public class Program
{
    public static void Main()
    {
        IContainer parent = new Container();
        IContainer child = parent.CreateChild();

        parent.Register<SomeType1>();
        child.RegisterDelegate<SomeType1>(_ => parent.Resolve<SomeType1>()); // very explicit about what you want to achieve

        Console.WriteLine("parent: " + parent.IsRegistered<SomeType1>()); // true 
        Console.WriteLine("child: " + child.IsRegistered<SomeType1>());   // true

        var x = child.Resolve<SomeType1>();
        Console.WriteLine(x);
    }

    class SomeType1 {}
}
```

**Comment by djymdmannen** [comment]
Ok, thanks! I was thinking of something along those lines as well. Need to do some testing and see. :)
